Fix #13486 (handle FillState when importing data from an old realm)

arpit551 opened the pull request:
Fixes [issue](https://github.com/zulip/zulip/issues/13486).

This PR is in continuation with [PR](https://github.com/zulip/zulip/pull/13608). First commit is reviewed there.

arpit551 commented:
@timabbott This is ready for review.

timabbott commented on zerver/lib/export.py:
```diff
@@ -305,16 +305,41 @@ def write_data_to_file(output_file: Path, data: Any) -> None:
+        if exclude and f.name in exclude:
+            continue
+        data[f.name] = f.value_from_object(instance)
+    return data
```
Can you explain precisely why you needed this modified `to_dict` field?  I see its intended to handle "non editable fields"; what non-editable fields are present in FillState?  A traceback for the error you're working around might be helpful.

timabbott replied:
(Also, a comment just before whichever line(s) we changed from the original implementation would be helpful, e.g. "# Zulip change: ...").

timabbott replied:
Aside from this question, the `FillState` export commit looks reasonable to me.

arpit551 replied:
@timabbott According to Django any field with auto_now=true will inherit editable=false too so in FillState model we have last_modified that has auto_now. When model_to_dict was used it had a check on non-editable fields. In to_dict function that check is removed. Although I tried to investigate why model_to_dict had that check to avoid non-editable field but got no result.

timabbott replied:
OK, interesting.  It's possible the right answer is that we should just remove the `last_modified` field; I'm not sure what it's used for.

timabbott replied:
I think the right answer here is to just remove that column as unused clutter rather than copying this code to work around it.

timabbott replied:
I just merged 9ac3e1099c60e7267a0fe307919b46916648c25d removing it.  Thanks for bringing this to my attention!

timabbott commented on analytics/lib/counts.py:
```diff
@@ -188,23 +188,36 @@ def do_aggregate_to_summary_table(stat: CountStat, end_time: datetime,
     end = time.time()
     logger.info("%s InstallationCount aggregation (%dms/%sr)" % (
-        stat.property, (end - start) * 1000, cursor.rowcount))
-    cursor.close()
```
It seems like removing this line is a bug; we should be closing the cursor when we're done with it.

timabbott commented on analytics/lib/counts.py:
```diff
@@ -188,23 +188,36 @@ def do_aggregate_to_summary_table(stat: CountStat, end_time: datetime,
+        else:
+            filters['value'] = value
+            InstallationCount.objects.create(**filters)
+            row_count = row_count + 1
```
Can you explain the reasoning behind this revised `InstallationCount` algorithm?  An alternative approach that we could take is just deleting and rebuilding the InstallationCount rows for the relevant window using the pre-existing code, for example; would that be cleaner?  If we did it inside a translation, other clients wouldn't experience any disruption from that flow.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+# we have to count stats just for the target Realm.
+def handle_fill_state(data: TableData, table: TableName, realm: Realm) -> None:
+    fill_state = FillState.objects.all()
+    if fill_state is None:
```
This is most clearly written `if fill_state.exists()`.

timabbott replied:
It may also be worth testing whether the Analytics code will avoid creating FillState` entries when run with an empty Zulip server (well, empty except for the system bots).  My guess is it won't, so this empty condition will only happen if one imports data into the server before the first `update_analytics_counts` cron run.  Maybe OK to leave this as a follow-up issue since it might be hard to fix without a lot of work, but we should at least document it with a comment.

arpit551 replied:
@timabbott Exactly so we have to define a condition where we can say that the server is truly empty.

timabbott replied:
OK.  I think that we should plan to write a function for that check that can live in the new library we'll be adding in https://github.com/zulip/zulip/issues/13736, and then this logic can use that.

I think it can be roughly checking that there's no realm with a name other than `zulipinternal` on the server; feel free to just add a function called `zulip_server_is_initialized_and_empty` at the time of `import_realm.py` doing that and then we can move it once #13736 is merged.

timabbott replied:
OK, we've finished the work required to unblock #13736.  If you're excited about unblocking this, you could claim that issue (I think the technical plan is pretty clear if you want to do it before Mateusz has a chance); it'd be a good opportunity to practice writing a super clear 10-commit series refactor.

arpit551 replied:
@timabbott Sure I will take a look at it.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+        # Assuming server is empty.
+        # TODO: A better condition is required here.
+        update_model_ids(FillState, data, 'analytics_fillstate')
+        bulk_import_model(data, FillState)
```
I would do an early return here, and then we don't need all the code below to be unnecessary indented in an `else` block.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+            items = check_imported_fill_state(items, stat, time_increment, realm)
+
+            if current_fill_state is None:
+                filters = {'property': stat.property, 'realm': realm}
```
Let's add a comment explaining the reasoning in each of these cases, and have them return.  E.g. this case is for when the server we exported from is ahead of the server we're importing into; our fix is to just delete the newer entries that we just imported.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+    COUNT_STATS = get_count_stats(filters['realm'])
+    stat = COUNT_STATS[filters['property']]
+    if isinstance(stat, LoggingCountStat) and stat.data_collector.output_table in [UserCount, StreamCount]:
+        RealmCount.objects.filter(**filters).delete()
```
Can you add a comment explaining the clearly the reasoning behind why this is the precise set of conditions when we only need to delete `RealmCount`?  

Also, I think it'd be better written as doing `RealmCount` unconditionally first, followed by an if statement for doing UserCount and StreamCount.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+
+def check_imported_fill_state(fill_state: Dict[str, Any], stat: CountStat,
+                              time_increment: timedelta, realm: Realm) -> Dict[str, Any]:
+    if fill_state['state'] is FillState.STARTED:
```
Please check for equality rather than using `is` to compare integers; it's more idiomatic Python

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+        RealmCount.objects.filter(**filters).delete()
+
+def check_imported_fill_state(fill_state: Dict[str, Any], stat: CountStat,
+                              time_increment: timedelta, realm: Realm) -> Dict[str, Any]:
```
This deserves a docstring explaining its purpose, namely to ensure that the FillState and analytics data we're working with is in a consistent state, and that we move back 1 time increment if it's FillState.STARTED.

timabbott commented on zerver/lib/import_realm.py:
```diff
@@ -1303,3 +1306,78 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+
+    logging.info('Successfully handled FillState')
+
+def delete_counts_by_filters(filters: Dict[str, Any]) -> None:
```
Have you thought about how this function will interact with racing against `update_analytics_counts` running?

One idea for how to do this better would be to filter the `zerver_realmcount` data objects before inserting them into the database, rather than inserting them into the database and then deleting extra rows.

timabbott replied:
I suppose one option for how we could manage this is having this block of `import_realm.py` share the `settings.ANALYTICS_LOCK_DIR` lock with `update_analytics_count.py`, and just wait until it can grab that lock before starting the "analytics import" part of its process.

arpit551 replied:
@timabbott interaction with running update_analytics_counts will definitely be an issue. The option you suggested seems legit to me about locking the process.

timabbott commented:
@arpit551 I think we should put this answer in a comment, but how is this intended to work for the Slack import case?  In that case, we have no imported `*count` or `fillstate` tables, and so we'll just need to run `process_count_stat` for the imported realm to populate those values.

timabbott commented:
OK, I've finished a first-pass review on this.  I think with fairly small changes, we should be in a position to merge at least the `export.py` commit to further shrink the unmerged part of this branch.  The last commit, for doing import, I think needs a lot more work, but it's definitely in a promising direction.  One thing we'll want to think about is how we can test whether we've done the race-handling code correctly.

arpit551 commented:
@timabbott  I have changed the requested changes kindly take a look at them.

timabbott commented:
CI is failing; let me know when you've fixed that (and removed the code related to `last_modified`).

arpit551 commented:
@timabbott I have fixed that but not modified installation count query.  I think we should have a discussion before rewriting it.

timabbott commented:
Great.  Happy to discuss; the main concern I have is that the current code in master uses nice bulk SQL queries that involve the work being done very efficiently in a single postgres transaction, and I'd like to avoid losing those nice properties while still preserving the correctness of the implementation.

arpit551 commented:
@timabbott How about using REPLACE rather than INSERT.
`    installationcount_query = """
        REPLACE INTO analytics_installationcount
            (value, property, subgroup, end_time)
        SELECT
            sum(value), '%(property)s', analytics_realmcount.subgroup, %%(end_time)s
        FROM analytics_realmcount
        WHERE
            property = '%(property)s' AND
            end_time = %%(end_time)s
        GROUP BY analytics_realmcount.subgroup
        """ % {'property': stat.property}`

arpit551 commented:
@timabbott How about using REPLACE rather than INSERT.
installationcount_query = """ REPLACE INTO analytics_installationcount (value, property, subgroup, end_time) SELECT sum(value), '%(property)s', analytics_realmcount.subgroup, %%(end_time)s FROM analytics_realmcount WHERE property = '%(property)s' AND end_time = %%(end_time)s GROUP BY analytics_realmcount.subgroup """ % {'property': stat.property}

arpit551 commented:
@timabbott How about using this.(I don't have much experience on  queries)
`INSERT INTO analytics_installationcount
            (value, property, subgroup, end_time)
        SELECT
            sum(value), '%(property)s', analytics_realmcount.subgroup, %%(end_time)s
        FROM analytics_realmcount
        WHERE
            property = '%(property)s' AND
            end_time = %%(end_time)s
        GROUP BY analytics_realmcount.subgroup
        ON CONFLICT (property, subgroup, end_time) DO UPDATE
        SET value = EXCLUDED.value`
Or we should leave it as it is for now and just delete and rebuild the InstallationCount rows for the relevant window at the time of processing counts for a particular realm(As done in the latest commit).

timabbott commented:
I really like that approach.  It just sounds like we need to fix https://github.com/zulip/zulip/issues/12362 first (aka the bug with the unique constraint).  I've been meaning to do the work to unblock upgrading to Django 2.x; so maybe let's divert to working on that and plan to come back to this?

timabbott commented:
See https://chat.zulip.org/#narrow/stream/3-backend/topic/django.20upgrade/near/812674 for a thread on Django upgrade work.  I think it's something we should be able to achieve in the next couple weeks; it might make sense for you to switch focus to another project while we work on that (or you can help out with the upgrade).  Let me know if you'd like pointers for something to work on.

arpit551 commented:
@timabbott I would love to contribute more to zulip. Let me know if you have something for me to work upon.

vinitS101 commented:
@arpit551 The [Issues Page](https://github.com/zulip/zulip/issues) has a number of open (and unclaimed) issues that you could work on.

timabbott commented:
We just merged the django 2.2 upgrade; can you submit a PR for https://github.com/zulip/zulip/issues/12362 next?  And then we can pick this back up once that's merged.

arpit551 commented:
@timabbott Sure I am on it

arpit551 committed:
Export: Exported FillState.

This commit includes adding a child to analytics_config to
export FillState table.

arpit551 committed:
server_initialization: Add zulip_server_isempty function.

Added zulip_server_isempty function that checks if any realm is
present in the server other than SYSTEM_BOT_REALM.

arpit551 committed:
analytics: Modified installation count query.

Add support for updating installation data after running for an
individual realm.
Modified test cases in test_counts.py to aggregate to InstallationCount
when there is a realm constraint.

arpit551 committed:
Import: Handled FillState while importing Realm.

FillState is handled according to the result of comparison between
old server and new server. Also its compatibility with running
update_count_stats is taken care by sharing the settings.ANALYTICS_LOCK_DIR
lock with update_analytics_count.py and wait until it grabs the lock.

Fixes #13486

arpit551 committed:
Import: Handled slack import case.

Count stats for imported realm where analytics data is
not available like in the case of slack import.

arpit551 commented:
@timabbott any more changes required?

arpit551 commented on zerver/lib/import_realm.py:
```diff
@@ -1315,3 +1326,126 @@ def import_analytics_data(realm: Realm, import_dir: Path) -> None:
+
+# Will wait until it grabs lock (settings.ANALYTICS_LOCK_DIR) from update_analytics_count.py
+# so that function below it will not create an issue with running update_analytics_counts
+def grab_analytics_lock() -> None:
```
I think this needs an alternative approach as it doesn't look much appealing.

shanukun commented on analytics/lib/counts.py:
```diff
@@ -193,28 +193,26 @@ def do_aggregate_to_summary_table(stat: CountStat, end_time: datetime,
+        ON CONFLICT
+            (property, subgroup, end_time)
+        DO UPDATE SET
+            value = EXCLUDED.value
```
@arpit551 This was the correct approach, but it doesn't work now that you'd fixed [#13486](https://github.com/zulip/zulip/issues/13486) in https://github.com/zulip/zulip/commit/f299f31340ee36c8af700f9715c7e3001eebfb98 (as there are now two unique constraints where `ON CONFLICT` can handle only one).
Do you have any other suggestions for this?

One approach which can help us fix this is to use `COALESCE`, you can read more about it here https://chat.zulip.org/#narrow/stream/49-development-help/topic/updating.20installation.20count.20table, do tell us what are your thoughts on this?
